### DISCUSSION TITLE
Issues at rendering to file (non=faster-than realtime)

### DISCUSSION THREAD

**Discussion by GeorgA** [savonet/liquidsoap#2814] [Q&A]
Due to the nice cue/fade-effects, I'm trying to use liquidsoap ( 2.1.4+git@5d5c1c71) as an "offline" renderer for audio files. It's based on a m3u with the files and an additional metadata file for cue-points etc. In principle it looks like that:

```
# read playlist once, call shutdown() when finished
input=playlist(playlistfile, mode="normal", loop=true, reload=1, reload_mode="rounds", on_reload=my_shutdown)

# ...cue/metadata/crossfade-stuff here (not relevant, issue occurs also without it)

clock.assign_new(sync="none",[input])    # force no-realtime

output.file(%mp3(bitrate=320),
	 fallible=true,
	 outputfile,
	input)
```
The issue is that this renders incomplete, actually it misses most of the files that are on a remote samba server. Frequently it also hangs at "Waiting for main threads to terminate" on shutdown. If I disable the clock.assign_new, it works as expected. The output matches exactly the playlist, so it's not a ffmpeg/decoding/fileaccess-problem. But it is rendered in realtime, which is not really acceptable. 

So it seems that there is a internal synchronization issue or glitch with the access/decoding times and the dataflow. If I cache the same files locally, there's also no issue, so I can use that as a workaround for now. But that gives me a slight uneasiness and I would like to avoid that :)

Now I wonder if this (non-streaming) use case is actually supported. Except the clock.assign_new-hack I didn't find any examples of similar rendering jobs.

**Comment by GeorgA** [comment]
Found a really really ugly workaround: I've added a on_track-handler to the playlist and sleep() a bit in that handler 8-O

Since the problem also reliably appears for local files if the log-output is reduced with log.level=1, I tried what sleep values work. 3ms doesn't, 5ms works most of the time. So I've set it to 100ms to be safe. But for reliability I still have to work with cached local files... 

But the main issue is still lurking inside the dataflow handling...

**Reply by toots** [reply]
Thanks for reporting. It's possible that the clock is performing too aggressively w.r.t. the samba asynchronous downloads. Those still need to happen and cannot be sped up but the general streaming clock will expect available requests too quickly.

We're planning a total rework of the clock system for the next major development cycle, this definitely will be a good use-case to keep in mind.

As a workaround, I'd suggest to download and store all your files locally before releasing the full power of the unsynced clock 🙂
